### PR TITLE
[API]maintaining cart after login and registration

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         |master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php [hunk 1]
```diff
@@ -32,14 +36,24 @@ final class RegisterShopUserHandler implements MessageHandlerInterface
     /** @var CustomerProviderInterface */
     private $customerProvider;
 
+    /** @var CartToUserAssignerInterface */
+    private $cartToUserAssignerInterface;
+
+    /** @var bool */
+    private $assignCartToRegisterUser;
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php:43]: This is just a proposal, but with this shop owner can change if he want registering user to has cart assigned or not with just parameter in services.xml 
I have made same to login in user.
Wdyt?

File: src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php [hunk 2]
```diff
@@ -95,6 +107,17 @@ public function iConfirmThisPassword(): void
...
 
+    /**
+     * @When I register with email :email and password :password
+     */
+    public function iRegisterWithEmailAndPassword(string $email, string $password): void
+    {
+        $this->fillContent($email, $password);
+        $this->iRegisterThisAccount();
+        $this->enableAccount($email);
+        $this->loginContext->iLogInAsWithPassword($email,$password);
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php:118]:
```suggestion
        $this->loginContext->iLogInAsWithPassword($email, $password);
```

File: src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php [hunk 3]
```diff
@@ -186,4 +225,28 @@ private function assertFieldValidationMessage(string $path, string $message): vo
...
     }
+
+    private function fillContent(?string $email = 'example@example.com', ?string $password = 'example'): void
+    {
+        $this->content = [
+            'firstName' => 'First',
+            'lastName' => 'Last',
+            'email' => $email,
+            'password' => $password
+            ];
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php:236]:
```suggestion
        ];
```

File: src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php [hunk 4]
```diff
@@ -186,4 +225,28 @@ private function assertFieldValidationMessage(string $path, string $message): vo
...
+    }
+
+    private function enableAccount(string $email): void
+    {
+        // Temporary implementation until we will have an endpoint to enable account by administrator
+        // 'on this channel account verification is not required' step from ChannelContext does not work
+        $repository = $this->shopUserManager->getRepository(ShopUser::class);
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php:243]: why didn't you use here `sylius.repository.user`?

File: src/Sylius/Bundle/ApiBundle/Assigner/CartToUserAssigner.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    public function __construct(ObjectManager $orderManager, SessionInterface $session, OrderRepositoryInterface $cartRepository){
+        $this->orderManager = $orderManager;
+        $this->session = $session;
+        $this->cartRepository = $cartRepository;
+    }
+
+    public function assignByCustomer(CustomerInterface $customer): void
+    {
+        if ($this->session->has('cart_token')) {
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Assigner/CartToUserAssigner.php:42]: shouldn't you cover here also case when `cart_token` will not be found? (some exception?)

File: src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php (hunk 1 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php:43]: IMHO it should be only one logic: after registration user has the previous cart. if someone wants to have a different 
cart flow, he should make customization with it.

File: src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php [hunk 6]
```diff
@@ -186,4 +225,28 @@ private function assertFieldValidationMessage(string $path, string $message): vo
...
+    {
+        $this->content = [
+            'firstName' => 'First',
+            'lastName' => 'Last',
+            'email' => $email,
+            'password' => $password
+            ];
+    }
+
+    private function enableAccount(string $email): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php:239]: I'm not convinced to this method, this is a logic that should be implemented in backend, not just in test behaviour. In the UI context, there is a listener (`UserRegistrationListener`) that automatically enables an account if account verification on a current channel is not required or sends an email. As long as we don't handle emails in API, I would be for enabling an account depending on channel in some listener or even in registration handler for now.

File: src/Sylius/Bundle/ApiBundle/Assigner/CartToUserAssigner.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    private $orderManager;
+
+    /** @var SessionInterface */
+    private $session;
+
+    /** @var OrderRepositoryInterface */
+    private $cartRepository;
+
+    public function __construct(ObjectManager $orderManager, SessionInterface $session, OrderRepositoryInterface $cartRepository){
+        $this->orderManager = $orderManager;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Assigner/CartToUserAssigner.php:35]:
```suggestion
    public function __construct(ObjectManager $orderManager, SessionInterface $session, OrderRepositoryInterface $cartRepository)
    {
        $this->orderManager = $orderManager;
```

File: src/Sylius/Bundle/ApiBundle/Assigner/CartToUserAssigner.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+namespace Sylius\Bundle\ApiBundle\Assigner;
+
+use Doctrine\Persistence\ObjectManager;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Sylius\Component\Customer\Model\CustomerInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+
+/** @experimental */
+class CartToUserAssigner implements CartToUserAssignerInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Assigner/CartToUserAssigner.php:23]: There is a `CartBlamerListener` that should work instead of this assigner I guess

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php [hunk 9]
```diff
@@ -98,6 +104,8 @@ public function __invoke(PickupCart $pickupCart)
 
         $this->orderManager->persist($cart);
 
+        $this->session->set('cart_token', $cart->getTokenValue());
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:107]: I guess that we shouldn't use session in handler

File: src/Sylius/Bundle/ApiBundle/Decorator/AuthenticationSuccessHandlerDecorator.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+
+
+/** @experimental */
+class AuthenticationSuccessHandlerDecorator implements AuthenticationSuccessHandlerInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Decorator/AuthenticationSuccessHandlerDecorator.php:31]: If I understood correctly, we are putting a cart to session during registration, but we are assigning it to the customer during login. What if after registration someone else will log in? 

I would say, that extending registration or login context, so the token can be sent explicitly to us would be a much better solution. Another option (but I think a little bit worse) would be to have custom cart claiming endpoint

File: src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php (hunk 1 above)
Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php]: Removed 😃

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php (hunk 9 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:107]: I could skip the conversation, but did we decided that we still want to use the session here? Is it used anywhere?

File: src/Sylius/Bundle/CoreBundle/EventListener/CartBlamerListener.php [hunk 11]
```diff
@@ -56,6 +57,16 @@ public function onInteractiveLogin(InteractiveLoginEvent $interactiveLoginEvent)
         $this->blame($user);
     }
 
+    public function onJWTAuthenticatedLogin(JWTAuthenticatedEvent $JWTAuthenticatedEvent): void
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/EventListener/CartBlamerListener.php:60]:
```suggestion
    public function onJWTAuthenticatedLogin(JWTAuthenticatedEvent $event): void
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php [hunk 12]
```diff
@@ -58,6 +66,18 @@ public function __invoke(RegisterShopUser $command): void
...
 
+        $this->handleVerification($user);
+
         $this->shopUserManager->persist($user);
     }
+
+    private function handleVerification(ShopUserInterface $user): void
+    {
+        /** @var ChannelInterface $channel */
+        $channel = $this->channelContext->getChannel();
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php:77]: In the long term, I would prefer to skip the usage of contexts in handlers. I think we should put data like channel into command.

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php (hunk 9 above)
Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:107]: Sorry, i must have forgote to delete this line 😃

File: src/Sylius/Bundle/CoreBundle/EventListener/CartBlamerListener.php (hunk 11 above)
Comment (arti0090) [reply src/Sylius/Bundle/CoreBundle/EventListener/CartBlamerListener.php:60]: I would stay with JWTAuthenticatedEvent to be consistent with rest of functions in this class

File: src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php (hunk 12 above)
Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php:77]: Hmm, do you have any resolution for this that we can implement now? Creating this function i was basing on ShopBundle->UserRegistrationListener. Should we create something similar or we leave it as it is and go back to it, when we make API user verification feauture? @lchrusciel

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php [hunk 13]
```diff
@@ -50,20 +51,25 @@ final class PickupCartHandler implements MessageHandlerInterface
     /** @var RandomnessGeneratorInterface */
     private $generator;
 
+    /** @var SessionInterface */
+    private $session;
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:56]: Adding session could be reverted as it is unused in this service

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RegisterShopUserHandlerSpec.php [hunk 14]
```diff
@@ -1,84 +0,0 @@
-<?php
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RegisterShopUserHandlerSpec.php:1]: Why did you remove this spec? 😱

File: src/Sylius/Bundle/CoreBundle/EventListener/CartBlamerListener.php (hunk 11 above)
Comment (GSadee) [reply src/Sylius/Bundle/CoreBundle/EventListener/CartBlamerListener.php:60]: I also vote for `event` here

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @arti0090! :tada:
